state/processor.go

```diff
@@ -49,27 +53,79 @@ type PseudoRandomizer interface {
 	Uint64() uint64
 }
 
+type StatePreImages struct {
+	rootHash common.Hash
+	preImages Transactions
+}
 
 type TransactionProcessor struct {
 	rand PseudoRandomizer
-	globalState GlobalStateDB
+	globalState *StateDB
+	prevStates map[LayerID]common.Hash
+	currentLayer LayerID
+	rootHash common.Hash
+	stateQueue list.List
+	db *trie.Database
+	mu sync.Mutex
 }
 
-func NewTransactionProcessor(rnd PseudoRandomizer, db GlobalStateDB) *TransactionProcessor{
+const maxPastStates = 20
+
+func NewTransactionProcessor(rnd PseudoRandomizer, db *StateDB) *TransactionProcessor{
 	return &TransactionProcessor{
 		rand: rnd,
 		globalState:db,
+		prevStates : make(map[LayerID]common.Hash),
+		currentLayer: 0,
+		rootHash: common.Hash{},
+		stateQueue: list.List{},
+		db : db.TrieDB(),
+		mu : sync.Mutex{},
 	}
 }
 
 //should receive sort predicate
-func (tp *TransactionProcessor) ApplyTransactions(transactions Transactions) error{
+func (tp *TransactionProcessor) ApplyTransactions(layer LayerID, transactions Transactions) error{
+	//todo: need to seed the mersenne twister with random beacon seed
+
 	txs := tp.mergeDoubles(transactions)
-	return tp.Process(tp.randomSort(txs), tp.coalesceTransactionsBySender(txs))
+	tp.Process(tp.randomSort(txs), tp.coalesceTransactionsBySender(txs))
+	tp.mu.Lock()
+	newHash , err := tp.globalState.Commit(false)
+	log.Info("new state root for layer %v is %x", layer, newHash)
+	tp.mu.Unlock()
+	if err != nil {
+		log.Error("db write error %v", err)
+		return err
+	}
+
+	tp.stateQueue.PushBack(newHash)
+	if tp.stateQueue.Len() > maxPastStates {
+		hash := tp.stateQueue.Remove(tp.stateQueue.Back())
+		tp.db.Commit(hash.(common.Hash),false)
+	}
+	tp.prevStates[layer] = newHash
+	tp.db.Reference(newHash, common.Hash{})
 	//	//call merge
 	//	//eliminate doubles
 	//	//check for double spends? (how do i do this? check the nonce against the prev one?
+	return nil
+}
 
+func (tp *TransactionProcessor) Reset(layer LayerID){
+	if state, ok := tp.prevStates[layer]; ok {
+		tp.mu.Lock()
+		defer tp.mu.Unlock()
+		newState, err := New(state, tp.globalState.db)
+		log.Info("reverted, new root %x", newState.IntermediateRoot(false))
+		if err != nil {
+			panic("wtf")
```

Comment:
why not return an error ?

Reply:
This is not something recoverable i think. anyway I'd like for this to be a panic for now, if we notice we get this one a lot during our testing we can think of why. but processing shouldn't continue after